### PR TITLE
BUD-107 & 127 -- User & Activity Caching

### PR DESCRIPTION
**Pull request by JakeThurman**
This API requires Idempotent callbacks.

**Idempotent**: 
> Idempotence is the property of certain operations in mathematics and computer science, that they can be applied multiple times without changing the result beyond the initial application.

i.e., Assume the callback given to `useUser` or `useActivity` could be called over and over with the same data. -- **Don't just append to a local list. Replace with the new data.**

### CODE REVIEW

**Comment by noahtallen** [comment]
With idempotent callbacks, are we really thinking about the underlying data source in the view as immutable data structures?

File: Buddies/Models/Activity.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        self.location = location
+        self.ownerId = ownerId
+        self.title = title
+        self.description = description
+        self.startTime = startTime
+        self.endTime = endTime
+        self.topicIds = topicIds
+    }
+    
+    static func from(snap: DocumentSnapshot, with delegate: ActivityInvalidationDelegate) -> Activity? {
```
Comment (noahtallen) [inline Buddies/Models/Activity.swift:70]: Could you explain the `ActivityInvalidationDelegate` a bit more?

File: Buddies/Models/User.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        didSet {
+            onChange("should_send_activity_suggestion_notification",
+                     shouldSendActivitySuggestionNotification)
+        }
+    }
+    
+    // MARK: Blocking
+    var blockedUsers : [UserId] { didSet { onChange("blocked_users", blockedUsers) } }
+    var blockedActivities : [ActivityId] { didSet { onChange("blocked_activities", blockedActivities) } }
+    let blockedBy : [UserId] // Shouldn't be updated directly! Automatic inverse of blocked_users.
```
Comment (noahtallen) [inline Buddies/Models/User.swift:53]: Should this property be `readonly` then?

File: Buddies/Utilities/DataAccessor.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+class DataAccessor : UserInvalidationDelegate, ActivityInvalidationDelegate {
+    static let instance = DataAccessor()
+    
+    var userListeners: [UserId : [Listener<User>]] = [:]
+    var activityListeners: [ActivityId : [Listener<Activity>]] = [:]
+    
+    func useUser(id: UserId, fn: @escaping (User) -> Void) -> (() -> Void) {
+        // TEMP
+        let testUser = User(delegate: self, imageUrl: "https://firebasestorage.googleapis.com/v0/b/beans-buddies.appspot.com/o/users%2F0VGQuKhyutT9YN12mobZauv6gtx2%2FprofilePicture.jpg?alt=media&token=1bc36e80-e3bb-4fbe-9444-50937651c3bd", isAdmin: true, uid: "n5vz1YkFqiP2IJ5rYDCEGve3QlG2", name: "Luke the Dummy", bio: "Hello blah blah blah bio", email: "noahtallen@waitImLuke.com", facebookId: nil, favoriteTopics: [], blockedUsers: [], blockedBy: [], blockedActivities: [], dateJoined: Date(), location: nil, shouldSendJoinedActivityNotification: true, shouldSendActivitySuggestionNotification: true, notificationToken: nil, chatReadAt: [:])
```
Comment (noahtallen) [inline Buddies/Utilities/DataAccessor.swift]: Where will this eventually come from?

File: Buddies/Utilities/DataAccessor.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if userListeners[id] == nil { userListeners[id] = [] }
+        userListeners[id]?.append(callback)
+        
+        // Call the function initially,
+        //  TODO: load actual user data
+        callback.fn(testUser)
+    
+        // Return a cancel callback
+        return {
+            self.userListeners[id] = self.userListeners[id]?.filter {
```
Comment (noahtallen) [inline Buddies/Utilities/DataAccessor.swift]: So this listens to an explicit user, not a collection of users?

File: Buddies/Onboarding/SignUpInfoVC.swift [hunk 5]
```diff
@@ -171,19 +171,28 @@ class SignUpInfoVC: LoginBase, UIImagePickerControllerDelegate, UINavigationCont
     
     @IBOutlet weak var bioText: UITextView!
     
+    func stringUntil(_ s: String?, _ c: Character) -> String? {
```
Comment (noahtallen) [inline Buddies/Onboarding/SignUpInfoVC.swift:175]: could you add a comment explaining this function?

Comment (noahtallen) [reply Buddies/Onboarding/SignUpInfoVC.swift:175]: I guess I get it - like "get the string up until the first time you see this character". Didn't see "get it" right away

**Review by noahtallen — APPROVED** [review]

File: Buddies/Models/User.swift (hunk 2 above)
Comment (JakeThurman) [reply Buddies/Models/User.swift:53]: it's a let idiot

**Pull request merged**